mono/utils/mono-os-mutex.h

```diff
@@ -310,51 +310,62 @@ WINBASEAPI BOOL WINAPI InitializeCriticalSectionEx(LPCRITICAL_SECTION lpCritical
 
 #endif /* defined __MINGW32__ && !defined __MINGW64_VERSION_MAJOR && (_WIN32_WINNT >= 0x0600) */
 
-typedef CRITICAL_SECTION mono_mutex_t;
+// FIXME mono_mutex_t and mono_mutex_recursive_t.
+typedef struct mono_mutex_t {
+	union {
+		CRITICAL_SECTION critical_section;
+		SRWLOCK srwlock;
+	};
+	gboolean recursive;
+} mono_mutex_t;
+
 typedef CONDITION_VARIABLE mono_cond_t;
 
 static inline void
 mono_os_mutex_init (mono_mutex_t *mutex)
 {
-	BOOL res;
-
-	res = InitializeCriticalSectionEx (mutex, 0, CRITICAL_SECTION_NO_DEBUG_INFO);
-	if (G_UNLIKELY (res == 0))
-		g_error ("%s: InitializeCriticalSectionEx failed with error %d", __func__, GetLastError ());
+	mutex->recursive = FALSE;
+	InitializeSRWLock (&mutex->srwlock);
 }
 
 static inline void
 mono_os_mutex_init_recursive (mono_mutex_t *mutex)
 {
-	BOOL res;
+	mutex->recursive = TRUE;
+	BOOL res = InitializeCriticalSectionEx (&mutex->critical_section, 0, CRITICAL_SECTION_NO_DEBUG_INFO);
 
-	res = InitializeCriticalSectionEx (mutex, 0, CRITICAL_SECTION_NO_DEBUG_INFO);
-	if (G_UNLIKELY (res == 0))
-		g_error ("%s: InitializeCriticalSectionEx failed with error %d", __func__, GetLastError ());
+	g_assertf (res, "%s: InitializeCriticalSectionEx failed with error %d", __func__, GetLastError ());
```

Comment:
``
if (!x)
 g_error(y)
```
is old
```
g_assert(x, y)
```

is new equivalent, except, right, I forgot `__func__` becomes redundant. I'll remove that.

Reply:
The non-Windows ones should be cleaned up, agreed.
`g_error` and `g_assertf` are equally fatal.
I'll PR cleaning up the others, or can restore these if you insist.

Reply:
They could have different behaviour since you can replace g_log_default_handler and using g_assertf will remove that capability, so changing away from g_error to g_assert should have more motivation than just old vs new (and I'm not aware that we have a decision moving in that direction).

Reply:
I can change it back, but really, `g_error` is fatal, presumed by many/all calls in the runtime to it. You cannot change that lightly or probably ever.

```
void
g_log_default_handler (const gchar *log_domain, GLogLevelFlags log_level, const gchar *message, gpointer unused_data)
{
	android_log (to_android_priority (log_level), log_domain, message);
	if (log_level & fatal)
		g_assert_abort ();
}
```

so `g_assertf` is just a nicer more compact form.